fix: ミドルウェアのリファクタ

shuta13 opened the pull request:
- #43

shzawa committed:
Reduction any (#53)

* add: MessageEventParam

* update: 不要なregexの削除

* add: MiddlewareParam

* add: getPostedURL()の引数に型指定

Co-authored-by: asato nago <35181442+75asa@users.noreply.github.com>

taichi0514 committed:
Feature/compile ci (#66)

* add compile_ci

* add husky

* fix ci yarn

shuta13 committed:
wip: replace enableAll

shuta13 committed:
feat: enableAll でまとめた

shuta13 committed:
fix: delete

shuta13 commented:
- バグってそうです
  - 投稿が6連チャンぐらい POST される

75asa commented on src/index.ts:
```diff
@@ -17,12 +17,9 @@ const app = new App({
-app.use(middleware.addUsersInfoContext);
-app.use(middleware.getFileInfo);
+(async () => {
+  await middleware.enableAll(app);
```
これって 一番下の即時関数に入れたらダメなの？

75asa reviewed (commented):
@shuta13 方向性あってます〜
いくつかコメント＆質問

shuta13 committed:
fix: bug of posting

tk-ozawa commented on src/customMiddleware.ts:
```diff
@@ -1,7 +1,7 @@
 import { MiddlewareParam } from "./bolt.interface";
 
-export const notBotMessages: any = async ({
+export const notBotMessages: (args: MiddlewareParam) => Promise<void> = async ({
   message,
   next,
 }: MiddlewareParam) => {
```
`const hoge: (args: Hogehoge) => Promise<void> = async () => {}`
について質問っす

- `(args: Hogehoge) => Promise<void>`の部分って何の型を指しているんですか？(右辺の`async () => {}`の戻り値の型?
  - argsと({ message, next })それぞれで引数を受け付けているのはなぜです？(メリット的なものも知りたい)

概念/単語レベルでわかんなくて調べられないのでお聞きしたいっす

shuta13 replied:
うおー、今見ました

> (args: Hogehoge) => Promise<void>の部分って何の型を指しているんですか？(右辺のasync () => {}の戻り値の型?

`(args: Hogehoge) => Promise<void>` は `notBotMessages` の型です。[Function 型](https://typescript-jp.gitbook.io/deep-dive/type-system/functions#no) とか言ったりするやつです(リンク先の `type ShortHand`)

`(args: Hogehoge) => Promise<void>` は`Hogehoge` 型の引数 `args` を受け取り、 `async` は `Promise` を暗黙的に返すので返り値の型は `Promise<void>` になります

もし `async` 関数の中で `return` していた場合は `Promise<return した値の型>` になります(知ってるかもしれないですが一応)

> argsと({ message, next })それぞれで引数を受け付けているのはなぜです？(メリット的なものも知りたい)

`({ message, next })` は引数として受け取ったオブジェクトから、特定の要素を取り出す書き方です。[分割代入](https://developer.mozilla.org/ja/docs/Web/JavaScript/Reference/Operators/Destructuring_assignment)とか言われるときがあります( `const { message, next } = args` と捉えると分割代入っぽく見えてくると思います )

もちろん `({ message, next })` は `(args)` にして `args.message` `args.next` で呼び出せます。

`({ message, next })` で書くメリット(メリットかどうかはわからないけど)は
- `console.log({ message })` みたいにログを吐かせられる
  - `console.log({ args.message })` は構文エラー
- https://qiita.com/amamamaou/items/1ec21316b8bf05ba9c34#%E5%BF%9C%E7%94%A8%E7%B7%A8 のような応用

こういう書き方もあるんだな〜程度でいいと思います。

tk-ozawa commented on src/customMiddleware.ts:
```diff
@@ -114,9 +115,18 @@ export const getChannelInfo: any = async ({
-export const enableAll: any = async (app: App): Promise<void> => {
+export const enableAll = async (app: App) => {
   if (process.env.SLACK_REQUEST_LOG_ENABLED === "1") {
     app.use(async (args) => {
```
ここの`args`って引数いきなり出てきたんですけどこれ is 何ですか?
それとこれは何が入っている感じです？(全くわからず。。)

75asa replied:
> ここの`args`って引数いきなり出てきたんですけどこれ is 何ですか?
> それとこれは何が入っている感じです？(全くわからず。。)

残余引数かな？
ちなみに、これはオリジナルの @seratch さんが書いてたコードです

seratch replied:
どの listener が呼ばれるか確定していない状態でとりうる middleware args すべてにマッチする可能性がある型がきます。それが、どんな型なのかは... 一言では言えませんね。こんな感じ... Bolt for JS の TS 型定義はちょっと複雑なので VS Code でホバーして型をチェックする方が良いと思います。

<img width="848" alt="Screen Shot 2020-12-14 at 15 03 42" src="https://user-images.githubusercontent.com/19658/102046078-ec0a3e00-3e1d-11eb-99ae-7bfc4322dfe2.png">

見ての通り、確実に存在するのは [AllMiddlewareArgs](https://github.com/slackapi/bolt-js/blob/%40slack/bolt%402.5.0/src/types/middleware.ts#L20-L26) だけ（これら以外に `body` なんかも必ず存在しますが）。普段は、この一つのオブジェクトから `({body, next}) =>` みたいにして、欲しいものだけを名前で取り出しているわけです。

75asa replied:
なるほど、 @seratch さんありがとうございます！ 🙇🏾 🙇🏾‍♀️ 🙏🏾

tk-ozawa replied:
@seratch @75asa ありがとうございます😌

75asa replied:
FYI [共変性と反変性 (計算機科学)](https://ja.wikipedia.org/wiki/%E5%85%B1%E5%A4%89%E6%80%A7%E3%81%A8%E5%8F%8D%E5%A4%89%E6%80%A7_(%E8%A8%88%E7%AE%97%E6%A9%9F%E7%A7%91%E5%AD%A6))

shzawa reviewed (commented):
@shuta13 @75asa 
質問したい箇所にコメント入れてるのでお手隙の際に対応願います🙇‍♂️

dependabot[bot] committed:
build(deps): bump ini from 1.3.5 to 1.3.8 (#71)

Bumps [ini](https://github.com/isaacs/ini) from 1.3.5 to 1.3.8.
- [Release notes](https://github.com/isaacs/ini/releases)
- [Commits](https://github.com/isaacs/ini/compare/v1.3.5...v1.3.8)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

shuta13 committed:
fix: refactor index.ts

shuta13 committed:
wip: reproduct bug

75asa committed:
fix: 🐛 compile.yml (#75)

75asa commented:
@shuta13 それと確認おなです 👩‍💻
master マージされた分が色々かわってるかと〜

75asa reviewed (commented):
@shuta13 色々conflict もしてるのでチェックオナです 🙏🏾

shuta13 committed:
chore: add comment

shuta13 commented on src/bolt.interface.ts:
```diff
@@ -0,0 +1,21 @@
+import { Context, MessageEvent, NextFn } from "@slack/bolt";
+import { WebClient } from "@slack/web-api";
+
+export interface MessageEventParam {
+  message: MessageEvent;
+  context: Context;
+  client?: WebClient;
+}
+
+export interface MiddlewareParam {
+  next?: NextFn;
+  client?: WebClient;
+  message?: MessageEvent;
+  context?: Context;
+}
+
+export interface SendMessageParam {
+  client: WebClient;
+  context: Context;
+  message: MessageEvent;
+}
```
最新の master を rebase やったんですがなんか特に conflict 起きてなかったぽいです...

75asa replied:
ファイル構成とかだいぶ変わってると思ったんだけど 🙄 
![image](https://user-images.githubusercontent.com/35181442/103129547-97e44100-46dc-11eb-80c1-e27f32d2c823.png)

75asa replied:
@shuta13 あ、 master じゃなくて `release/beta-2.0` にしてます！

75asa committed:
Chore/add lang ja issue template 2 (#82)

* chore: 🤖 add template ja_JP

* fix #67

75asa committed:
feat: 🎸 add basic template configrations (#84)

75asa committed:
docs: ✏️ update readme & add reelease npm script (#85)

75asa committed:
Chore/make issue adding project (#86)

* 2.0.0

* fix: 🐛 issue/config.yml

75asa committed:
chore: 🤖 update .gitignore & README.md (#87)

* chore: 🤖 update .gitignore & README.md

* fix #79

75asa committed:
Fix/issue template on automation projects (#88)

* fix: 🐛 issue config

* fix: 🐛 PR template name

* fix #89

75asa committed:
fix: 🐛 config.yml (#90)

75asa committed:
refactor: 💡 config & template (#91)

75asa committed:
Chore/update GitHub templates (#92)

* refactor: 💡 issue template

* refactor: 💡 PR template

75asa committed:
fix: 🐛 issues/config.yml (#97)

75asa commented on src/index.ts:
```diff
@@ -24,59 +24,22 @@ const sendMessage: (args: SendMessageParam) => Promise<void> = async ({
-  };
 
-  console.log("1回目", JSON.stringify(msgOption, null, 4));
+  if ("profile" in context && "channel" in context) {
```
`console.time` を使って各処理を計測してみました
middleware 全体の経過時間は `enableAll: 1053.091ms` と1秒ぐらい、メッセージ送信は `1051.180ms` と同じく1秒ぐらい
ただ、ミドルウェア全ての処理が終わるよりも前にメッセージ送信の方が優先されていることが謎ですね
以下の書き方ではだめなのか...🤔 

```src/index.ts
(async () => {
  // Enable All Middlewares
  await middleware.enableAll(app);

  // Send Message
  await app.message(sendMessage);

  // Start your app
  await app.start(process.env.PORT || 3000);

  console.info("⚡️ Bolt app is running!");
})();
```

```console
⚡️ Bolt app is running!
[DEBUG]  web-api:WebClient:0 http response received
[DEBUG]  web-api:WebClient:0 http response received
[DEBUG]  bolt-app ack() begin
[DEBUG]  bolt-app ack() response sent
[DEBUG]  web-api:WebClient:1 initialized
[DEBUG]  bolt-app Conversation context failed loading for ID: CTDKZ45V1, error: Conversation not found
[DEBUG]  web-api:WebClient:1 apiCall('team.info') start
[DEBUG]  web-api:WebClient:1 will perform http request
getFileInfo: 0.176ms
[DEBUG]  web-api:WebClient:1 apiCall('users.info') start
[DEBUG]  web-api:WebClient:1 will perform http request
notBotMessages: 0.012ms
noThreadMessages: 0.004ms
[DEBUG]  web-api:WebClient:1 apiCall('conversations.info') start
[DEBUG]  web-api:WebClient:1 will perform http request
(node:61591) Warning: Label 'sendMessage' already exists for console.time()
(node:61591) Warning: Label 'sendMessage' already exists for console.time()
(node:61591) Warning: No such label 'notBotMessages' for console.timeEnd()
[DEBUG]  web-api:WebClient:1 http response received
getTeamInfo: 416.475ms
(node:61591) Warning: Label 'sendMessage' already exists for console.time()
[DEBUG]  web-api:WebClient:1 http response received
addUsersInfoContext: 416.665ms
(node:61591) Warning: Label 'sendMessage' already exists for console.time()
[DEBUG]  web-api:WebClient:1 http response received
getChannelInfo: 588.856ms
1回目
[DEBUG]  web-api:WebClient:0 apiCall('chat.postMessage') start
[DEBUG]  web-api:WebClient:0 will perform http request
(node:61591) Warning: Label 'sendMessage' already exists for console.time()
[DEBUG]  web-api:WebClient:0 http response received
msg: ok ✅
sendMessage: 1051.180ms
enableAll: 1053.091ms
finished enable all
1回目
[DEBUG]  web-api:WebClient:0 apiCall('chat.postMessage') start
[DEBUG]  web-api:WebClient:0 will perform http request
[DEBUG]  web-api:WebClient:0 http response received
msg: ok ✅
sendMessage: 446.018ms
[DEBUG]  bolt-app next() call completed
[DEBUG]  bolt-app ack() begin
[DEBUG]  bolt-app ack() response sent
[DEBUG]  bolt-app ack() begin
[DEBUG]  bolt-app ack() response sent
```

`[DEBUG]  bolt-app Conversation context failed loading for ID: CTDKZ45V1, error: Conversation not found` となってるのも気になる
GUI の[conversations.info](https://api.slack.com/methods/conversations.info) の Tester でトークンと ID: `CTDKZ45V1` を入れて試したところ Payload がしっかり出ました 🤔

75asa replied:
@seratch さん、bolt-js で複数のミドルウェアを `enableAll()` などで使いたい場合は
https://github.com/seratch/deepl-for-slack/blob/master/src/custom-middleware.ts#L5 などのようにして
`app.use(` 内側に `Promise.all([middle_1, middle_2...)` などのようにして実装するのがベターなのでしょうか？

FYI
```
    app.use(async (args) => {
      console.time("enableAll");
      await Promise.all([
        getTeamInfo(args as MiddlewareParam),
        getFileInfo(args as MiddlewareParam),
        addUsersInfoContext(args as MiddlewareParam),
        notBotMessages(args as MiddlewareParam),
        noThreadMessages(args as MiddlewareParam),
        getChannelInfo(args as MiddlewareParam),
      ]);
      console.timeEnd("enableAll");
```

bolt-js 内部での処理として `app.use()` は以下のように定義されているかと思います
args の m を与えた場合、内部のミドルウェア処理にプッシュされているのは確認できたのですが、実装方法として推奨されているのかが不安になりました
（他の bolt を使用した複数ミドルウェア処理のコード例も発見できず...）

75asa replied:
@seratch さん お手隙の際に FB いただけると幸いです 🙏🏾  🙇🏾  🙇🏾‍♀️

seratch replied:
@shuta13 さすがにコードの詳細までフィードバックはできないですが、タイムアウトで再送が発生していないのなら subtype がついている message イベントに反応しているのかもしれません。profile の有無ではなく subtype なし or subtype が bot_message のものだけ反応するようにしてみてはどうでしょうか。

@75asa 

>[DEBUG] bolt-app Conversation context failed loading for ID: CTDKZ45V1, error: Conversation not found

これは、気にしなくていいです。紛らわしいからエラーとして出力しないようにしようという話をしていたりします `https://github.com/slackapi/bolt-js/issues/629`

>app.use( 内側に Promise.all([middle_1, middle_2...) などのようにして実装するのがベターなのでしょうか？

いや、普通にそれぞれ app.use で register すればいいんじゃないでしょうか。一つの middleware にしてその中でなんかパフォーマンス最適化したいとかそういうことのであれば、それはそれでよいとは思いますが。

（最初にメンションに反応しはしましたが、、私このプロジェクトのメンバーじゃないんでw あんまりメンションしないでくれると嬉しいです

75asa replied:
お忙しい中ありがとうございます！！
その方向で試してみます！
（すみません…以後ないようにします🙏）

75asa replied:
TODO: middleware の subtype を Logging

dependabot[bot] committed:
build(deps): bump axios from 0.19.1 to 0.21.1 (#98)

Bumps [axios](https://github.com/axios/axios) from 0.19.1 to 0.21.1.
- [Release notes](https://github.com/axios/axios/releases)
- [Changelog](https://github.com/axios/axios/blob/v0.21.1/CHANGELOG.md)
- [Commits](https://github.com/axios/axios/compare/0.19.1...v0.21.1)

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
